src/lib/lore.sh

```diff
@@ -458,3 +458,41 @@ function download_series()
   done
   wait
 }
+
+function apply_patch()
+{
+  local total_patches="$1"
+  local first_message_id="$2"
+  local patch_title="$3"
+  local kernel_tree="$4"
+  local target_branch="$5"
+  local flag="$6"
+  local path_to_tmp_dir
+  local filename_pattern
+  local files_to_remove
+  local cmd
+  local ret
+
+  path_to_tmp_dir=$(mktemp --directory)
+
+  download_series "${total_patches}" "${patch_url}" "${path_to_tmp_dir}" "${patch_title}"
+
+  cmd="git -C ${kernel_tree} switch ${target_branch}"
+  cmd_manager "$flag" "$cmd"
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    complain "Could not switch kernel tree ${kernel_tree} to branch ${target_branch}"
+    return "$ret"
+  fi
+
+  filename_pattern=$(convert_title_to_patch_name "${patch_title}")
+  files_to_apply=$(find "${path_to_tmp_dir}" -iname "*${filename_pattern}" | sort)
+  while IFS=' ' read -r file; do
+    cmd="git -C ${kernel_tree} am ${file}"
```

Comment:
Since `git am` can fail, we could use `--reject` to generate the `.rej` file and allow the developer to fix the issue manually. We can try our best to make this `am` very robust. One idea is to use the `wiggle` in case of git am failure for trying to apply the failed patch again.

We should update the target branch before trying to merge it. This can increase the chance of successfully applying a patch.

Another thing to consider is to create a new branch based on the `target_branch` to apply the new patch. We don't want to 'contaminate' the main branch with unmerged patches. You can create a pattern for those branches, maybe something like:

```
YYYY-MM-DD-HH-MM-SS-PATCH-TITLE-SHORT-SHA
```

Remember that applying patches can easily fail, and we must define some strategies. We could create a status system in the updtream-patches-ui to track if a patch failed to be applied.

1. https://linux.die.net/man/1/wiggle